### PR TITLE
Fix 2FA

### LINKED ISSUE #1964: Two-factor authentication crashing on Windows (RC 1.2)
### Before creating an issue, please ensure:
- [x] This is a bug in the software that resides in this repository, and not a
    support matter (use https://otland.net/forums/support.16/ for support)
- [x] This issue is reproducible without changes to the code in this repository
### Steps to reproduce (include any configuration/script required to reproduce)
1. Create account (username: testing1, password: testing1, secret: [OHWC267D3OQKDMTG](http://1.1m.yt/0B1cg-.png))
2. Login with correct username, password but invalid secret key, etc 123123
### Expected behaviour
1. Login should fail, user should be asked to write correct secret key
### Actual behaviour
1. Server crashes. 

~~if I write the correct token generated by "Authy" mobile app, it works and I can login just fine.~~
### Environment

Windows 10, 64bit. 
Tried to run 32bit and 64bit precompiled TFS 1.2 release versions.

Not quite sure what im doing but I managed to get gdb on windows through mingw32, here is what I was able to capture through debugging of the 32bit executable file:

```
C:\tfs13>gdb theforgottenserver.exe
GNU gdb (GDB) 7.6.1
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "mingw32".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from C:\tfs13\theforgottenserver.exe...(no debugging symbols found)...done.
```

```
(gdb) run
Starting program: C:\tfs13/theforgottenserver.exe
[New Thread 8032.0x2688]
[New Thread 8032.0x71c]
[New Thread 8032.0x27bc]
[New Thread 8032.0x2f70]
[New Thread 8032.0x2fa4]
[New Thread 8032.0x2d68]
[New Thread 8032.0x19f4]
```

```
The Forgotten Server - Version 1.2
Compiled with Microsoft Visual C++ version 14.0
Compiled on Oct 22 2016 03:34:15 for platform x86

A server developed by Mark Samman
Visit our forum for updates, support, and resources: http://otland.net/.

>> Loading config
>> Establishing database connection... MySQL 6.1.6
>> Running database manager
[New Thread 8032.0x185c]
>> Loading vocations
>> Loading items
>> Loading script systems
>> Loading monsters
>> Loading outfits
>> Checking world type... PVP
>> Loading map
> Map size: 1000x1000.
> Map loading time: 1.295 seconds.
>> Initializing gamestate
>> Loaded all modules, server starting up...
>> Forgotten Server Online!
```

```
gdb: unknown target exception 0xc0000409 at 0x751491cb

Program received signal ?, Unknown signal.
[Switching to Thread 8032.0x2d68]
0x751491cb in ucrtbase!abort () from C:\WINDOWS\SysWoW64\ucrtbase.dll
```

```
(gdb) frame 0
#0  0x751491cb in ucrtbase!abort () from C:\WINDOWS\SysWoW64\ucrtbase.dll
(gdb) frame 1
#1  0x75148abb in ucrtbase!terminate () from C:\WINDOWS\SysWoW64\ucrtbase.dll
(gdb) frame 2
#2  0x4cd6c780 in ?? ()
(gdb) frame 3
#3  0x0100e378 in ?? ()
(gdb) frame 4
#4  0x73902c87 in wcsstr () from C:\WINDOWS\SysWoW64\vcruntime140.dll
(gdb) frame 5
#5  0x73902ba8 in wcsstr () from C:\WINDOWS\SysWoW64\vcruntime140.dll
(gdb) frame 6
#6  0x7390e436 in _CxxFrameHandler3 ()
   from C:\WINDOWS\SysWoW64\vcruntime140.dll
```

```
(gdb) thread 1
[Switching to thread 1 (Thread 8032.0x2688)]
#0  0x7720dc7c in ?? ()
(gdb) thread 2
[Switching to thread 2 (Thread 8032.0x71c)]
#0  0x7720f7fc in ?? ()
(gdb) thread 3
[Switching to thread 3 (Thread 8032.0x27bc)]
#0  0x7720f7fc in ?? ()
(gdb) thread 4
[Switching to thread 4 (Thread 8032.0x2f70)]
#0  0x7720f7fc in ?? ()
(gdb) thread 5
[Switching to thread 5 (Thread 8032.0x2fa4)]
#0  0x7720dc2c in ?? ()
(gdb) thread 6
[Switching to thread 6 (Thread 8032.0x2d68)]
#0  0x751491cb in ucrtbase!abort () from C:\WINDOWS\SysWoW64\ucrtbase.dll
```

```
(gdb) bt full
#0  0x7720f7fc in ?? ()
No symbol table info available.
#1  0x748462c4 in KERNEL32!BaseThreadInitThunk ()
   from C:\WINDOWS\SysWoW64\kernel32.dll
No symbol table info available.
#2  0x77200609 in ?? ()
No symbol table info available.
#3  0x772005d4 in ?? ()
No symbol table info available.
#4  0x00000000 in ?? ()
No symbol table info available.
```

### PR DESCRIPTION
**Pull request by Mkalo**
This should fix #1964.

The crash was happening in:
```cpp
uint32_t truncHash = std::stol(message.substr(2 * offset, 8), nullptr,
16) & 0x7FFFFFFF;
```

For some reason it only crashed in Windows. I've changed all std::stol
to std::strtoul to prevent exceptions.

### CODE REVIEW

File: src/protocolgame.cpp [hunk 1]
```diff
@@ -281,7 +281,7 @@ void ProtocolGame::onRecvFirstMessage(NetworkMessage& msg)
 	std::string& accountName = sessionArgs[0];
 	std::string& password = sessionArgs[1];
 	std::string& token = sessionArgs[2];
-	uint32_t tokenTime = strtoul(sessionArgs[3].c_str(), NULL, 10);
+	uint32_t tokenTime = std::strtoul(sessionArgs[3].c_str(), NULL, 10);
```
Comment (djarek) [inline src/protocolgame.cpp]: Why aren't we checking for overflow in security related code?

Comment (ranisalt) [inline src/protocolgame.cpp]: Also doesn't strtoul receives a string as parameter?

Comment (Mkalo) [inline src/protocolgame.cpp]: It receives const char*

Comment (ranisalt) [inline src/protocolgame.cpp]: Then you should use `std::stoul(sessionArgs[3])`.

**Comment by gugahoa** [comment]
@ranisalt @Mkalo pr status?

**Comment by ranisalt** [comment]
Still waiting a change.

**Comment by ranisalt** [comment]
An issue should be opened in whatever tracker exists for Windows' equivalent of libc. It is a clear issue with `std::stoul` if `std::strtoul` works with the same C string.

**Comment by djarek** [comment]
It's not a bug it's a feature. `std::stoul` throws an exception on overflow, but `std::strtoul` sets errno (which we ignore).

**Comment by ranisalt** [comment]
How can a string that is at most 8 hex digits overflow when casted to int?

**Pull request merged**